feat: rework gateway page and add recursive gateway (#1966)
ipfs/ipfs-docs: docs/concepts/ipfs-gateway.md

Non-recursive gateways are gateways that only serve content that they have themselves. For example, [Kubo](../install/command-line.md) can be configured to act as a non-recursive gateway by setting the [`Gateway.NoFetch=true`](https://github.com/ipfs/kubo/blob/master/docs/config.md#gatewaynofetch) option.

In general, recursive gateways are more powerful for end-users because they abstract away all details of the peer-to-peer network. However, they are much more resource-intensive for operators and prone to abuse.

[Trustless, verifiable retrieval](../reference/http/gateway.md#trustless-verifiable-retrieval) from non-recursive gateways is becoming a popular way to provide IPFS content to the network ([HTTP](https://docs.ipfs.tech/reference/http/gateway/#trustless-verifiable-retrieval) as an alternative or in addition to [Bitswap](../concepts/glossary.md#bitswap)).

## Trusted vs. trustless gateways

> See [Trusted vs. Trustless Gateways](../reference/http/gateway/#trusted-vs-trustless) for more information.

Reviewer: ```suggestion
See [Trusted vs. Trustless Gateways](../reference/http/gateway.md#trusted-vs-trustless) for more information.
```

  - [Authenticated gateways](#authenticated-gateways)
- [Gateway request lifecycle](#gateway-request-lifecycle)
- [Resolution styles](#resolution-styles)
  - [Path](#path)
  - [Subdomain](#subdomain)
  - [DNSLink](#dnslink)
- [Gateway URL formats](#gateway-url-formats)
- [Working with gateways](#working-with-gateways)
- [Implementing gateways](#implementing-gateways)
- [Learning more](#learning-more)

## Gateway providers

Regardless of who deploys a gateway and where, any IPFS gateway resolves access to any requested IPFS [content identifier](content-addressing.md).

### Your local gateway

Your machine may host a gateway as a local service; e.g., at `localhost:8080`. You have a local gateway service if you installed [IPFS Desktop](https://github.com/ipfs-shipyard/ipfs-desktop#ipfs-desktop) or another form of IPFS node.

### Public gateways

Public ([recursive](#recursive-vs-non-recursive-gateways)) gateways are provided by the IPFS various other organizations, including the [IPFS Foundation](./public-utilities.md#public-ipfs-gateways) as a public utility.

For a list of public gateways, see the [IPFS Gateways Checker](https://ipfs.fyi/gateways).

> ## Gateway types

Author: I wonder if it makes sense to add a "Gateway providers" type for non-recursive gateways

In general, recursive gateways are more powerful for end-users because they abstract away all details of the peer-to-peer network. However, they are much more resource-intensive for operators and prone to abuse.

Non-recursive gateways are becoming a popular way to provide IPFS content to the network (as an alternative or in addition to Bitswap). Non-recursive provider records can only be announced to the [IPNI](../concepts/ipni/), but not the [DHT](../concepts/dht.md).

## Trusted vs. trustless gateways

See [Trusted vs. Trustless Gateways](../reference/http/gateway/#trusted-vs-trustless) for more information.

### Read-only gateways

_Read-only gateways_ are the simplest kind of gateway. This gateway type provides a way to fetch IPFS content using the HTTP GET method.

## Authenticated gateways

If a gateway provider wants to limit access to requests with authentication, they may need to configure a reverse proxy, develop an IPFS plugin, or set a cache-layer above IPFS.

Configuring a reverse proxy is the most popular way for providers handling authentication. Reverse proxy can also keep the original IPFS API calls which makes gateway adaptable to all IPFS SDK and toolkits.

> ![Auth with Reverse proxy](./images/ipfs-gateways/public-authed-gateway.png)

Reviewer: nit: this image is wrong and confusing (gateway has no `/api/v0`), just remove it.

```suggestion
```

## Gateway request lifecycle

:::callout

> This section uses the _default_ recursive gateway request lifecycle of [IPFS Kubo](https://github.com/ipfs/kubo) to introduce the basic concepts in the lifecycle. However, non-recursive gateways only serve content that they have and/or want to provide. For example, a Kubo gateway with `NoFetch` enabled will **not** attempt to retrieve content from the network.

Reviewer: ```suggestion
This section uses the _default_ recursive gateway request lifecycle of [IPFS Kubo](https://github.com/ipfs/kubo) to introduce the basic concepts in the lifecycle. However, non-recursive gateways only serve content that they have and/or want to provide. For example, a Kubo gateway with [`Gateway.NoFetch=true`](https://github.com/ipfs/kubo/blob/master/docs/config.md#gatewaynofetch) will **not** attempt to retrieve content from the network.
```

:::

When a client request for a CID reaches an IPFS gateway, the gateway first checks whether the CID is cached locally. At this point, one of the following occurs:

- **If the CID is cached locally**, the gateway responds with the content referred to by the CID, and the lifecycle is complete.

> - **If the CID is not in the local cache**, the gateway will attempt to retrieve it from the network.

Reviewer: Good opportunity to show difference between both gw types?

```suggestion
- **If the CID is not in the local cache**, a non-recursive gateway would error, however our gateway is recursive and will attempt to retrieve it from the network.
```

   https://{gateway URL}/ipns/{example.com}/{optional path}
   ```

> 2. The gateway searches the DNS TXT records of the requested domain `{_dnslink.example.com}` for a string of the form `dnslink=/ipfs/{CID}`. If found, the gateway uses the specified CID to serve up `ipfs://{CID}/{optional path}`. As with path resolution, this form of DNSLink resolution violates the single-origin policy. The domain operator may ensure single-origin policy compliance — and the delivery of the current version of content — by adding an `Alias` record in the DNS that refers to a suitable IPFS gateway; e.g., `gateway.ipfs.io`.
> 3. The `Alias` record redirects any access to that `example.com` to the specified gateway. Hence the browser's request to `https://{example.com}/{optional path to resource}` redirects to the gateway specified in the `Alias`.
> 4. The gateway employs DNSLink resolution to return the current content version from IPFS.
> 5. The browser does not perceive the gateway as the origin of the content and therefore enforces the single-origin policy to protect `example.com`.

Reviewer: Wow.. there is a lot of.. lets say untrue, controversial or vague statements in the old text here:
- We describe DNSLink name support on path gateway mixing it with logic and advanced setup based on `ALIAS` DNS record, and we are not mentioning `Host` header  case which is literally the core of https://specs.ipfs.tech/http-gateways/dnslink-gateway/
- Using alias is a gigantic footgun if you dont own both domains. It delegates authority to somneone, they can pretend to own your domain now. This is why Fleek uses CNAME for `_dnslink` subdomain, and nothing more.
- comment about origin isolation is incorrect (there is no isolation on path gateway, we protect noone)

I suspect this was glue together from various notes that made sense on their own, but in this form it is plain wrong.

Let's clean this up:

````suggestion
2. The gateway searches the DNS TXT records  on `_dnslink.` subdomain (`_dnslink.example.com`) for a string of the form `dnslink=/ipfs/{CID}`. If found, the gateway uses the specified content identifier to find and serve up `ipfs://{CID}/{optional path}`. 

It is possible to use HTTP gateway for serving content on the DNSLink domain itself:

1. Point `example.com` at IP of your HTTP gateway, make sure `A`/`AAAA`/`HTTPS` records are set, and TLS termination is configured.
2. Client sends request to:

   ```bash
   https://{example.com}/{optional path}
   ```

3. Gateway detects HTTP header `Host: example.com` in the incoming request and searches DNSLink the same way as in previous example.
````
